pypesto/objective/objective.py

```diff
@@ -0,0 +1,910 @@
+
+        self.index = index
+
+    def update(self, x, sensi_orders, mode, result):
+        """
+        Update the history.
+
+        Parameters
+        ----------
+        x: np.ndarray
+            The current parameter.
+        sensi_orders: tuple
+            The sensitivity orders.
+        mode: str
+            As in Objective.MODE_.
+        result: dict
+            The result for x.
+
+        Notes
+        -----
+
+        It is assumed that the result dictionary only contains those
+        values that were requested in Objective.___call__.
+        """
+        self._update_counts(sensi_orders, mode)
+        self._update_trace(x, mode, result)
+        self._update_vals(x, sensi_orders, mode, result)
+
+    def finalize(self):
+        """
+        Save the trace to file if options.trace_save is True.
+        """
+        self._save_trace(finalize=True)
+
+    def _update_counts(self, sensi_orders, mode):
+        """
+        Update the counters.
+        """
+        if mode == Objective.MODE_FUN:
+            if 0 in sensi_orders:
+                self.n_fval += 1
+            if 1 in sensi_orders:
+                self.n_grad += 1
+            if 2 in sensi_orders:
+                self.n_hess += 1
+        elif mode == Objective.MODE_RES:
+            if 0 in sensi_orders:
+                self.n_res += 1
+            if 1 in sensi_orders:
+                self.n_sres += 1
+
+    def _update_trace(self, x, mode, result):
+        """
+        Update and possibly store the trace.
+        """
+
+        if not self.options.trace_record:
+            return
+
+        # init trace
+        if self.trace is None:
+            columns = ['time',
+                       'n_fval', 'n_grad', 'n_hess', 'n_res', 'n_sres',
+                       'fval', 'grad', 'hess', 'res', 'sres',
+                       'x']
+            self.trace = pd.DataFrame(columns=columns)
+
+        # extract function values
+        if mode == Objective.MODE_FUN:
+            fval = result.get(Objective.FVAL, None)
+            grad = result.get(Objective.GRAD, None)
+            hess = None if self.options.trace_record_hess \
+                else result.get(Objective.HESS, None)
+            res = None
+            sres = None
+        else:  # mode == Objective.MODE_RES
+            res = result.get(Objective.RES, None)
+            sres = result.get(Objective.SRES, None)
+            # can compute fval from res
+            fval = res_to_fval(res)
+            # grad could also be computed from sres
+            grad = None
+            hess = None
+
+        used_time = time.time() - self.start_time
+
+        # create table row
+        values = [
+            used_time,
+            self.n_fval, self.n_grad, self.n_hess, self.n_res, self.n_sres,
+            fval, grad, hess, res, sres,
+            x
+        ]
+
+        # check whether to append to trace
+        if not self.options.trace_all and fval >= self.fval_min:
```

Comment:
we could check for that a bit earlier, right?

Reply:
At this point one remark: I found that pd.DataFrame.toc appending is really slow compared to list.append() (like 10s to <1s), but of course has better readability. Usually won't matter for our ODE problems.